### ISSUE TITLE
pos_pair_ = pos_pair_[pos_pair_ < 1 - epsilon] ?

### ISSUE THREAD

**Issue by raymongL** [MalongTech/research-ms-loss#14]
Hi there, thanks for sharing the code and beautifule work! 
In multi_similarity_loss.py line 35 ：
` pos_pair_ = pos_pair_[pos_pair_ < 1 - epsilon] `
why do we need this code ? 
And what's the logic of using the output of avgpooling as the embeddings of network?

**Comment by Bigwode** [comment]
A trick to exclude the similarity with itself, because the cosine similarity between one feature and itself is 1.

**Comment by raymongL** [comment]
Umm, but in line multi_similarity_loss.py line 28:
`sim_mat = torch.matmul(feats, torch.t(feats))`
you are using the dot product similarity?

**Comment by MarStarck** [comment]
so the feature must be normalized, right?

**Comment by MarStarck** [comment]
but I wonder that the input of loss should be logits, to compare with labels. so do I need to do a softmax first to make it between 0~1?

**Comment by shdeldari** [comment]
Hi Thank you for sharing the code. Could you please let me know which part is related to pair mining step? In the for loop in MultiSimilarityLoss.py file it is written:
for i in range(batch_size):
            pos_pair_ = sim_mat[i][labels == labels[i]]

Does it mean you know the labels? so you skip pair mining? 
I will really appreciate your help.
